java-lib/src/main/java/com/wavefront/ingester/Ingester.java

```diff
@@ -31,76 +33,92 @@
-          .option(ChannelOption.SO_BACKLOG, 1024)
-          .localAddress(listeningPort)
-          .childHandler(new ChannelInitializer<SocketChannel>() {
-            @Override
-            public void initChannel(SocketChannel ch) throws Exception {
-              ChannelPipeline pipeline = ch.pipeline();
-              pipeline.addLast(new LineBasedFrameDecoder(4096, true, true));
-              pipeline.addLast(new StringDecoder(Charsets.UTF_8));
-              if (decoders != null) {
-                for (Function<SocketChannel, ChannelHandler> handler : decoders) {
-                  pipeline.addLast(handler.apply(ch));
-                }
-              }
-              // Shared across all reports for proper batching
-              pipeline.addLast("idleStateHandler", new IdleStateHandler(CHANNEL_IDLE_TIMEOUT_IN_SECS,
-                  0, 0));
-              pipeline.addLast("idleChannelTerminator", new ChannelDuplexHandler() {
-                @Override
-                public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
-                  if (evt instanceof IdleStateEvent) {
-                    if (((IdleStateEvent) evt).state() == IdleState.READER_IDLE) {
-                      logger.warning("terminating connection to graphite client due to inactivity after " +
-                          CHANNEL_IDLE_TIMEOUT_IN_SECS + "s: " + ctx.channel());
-                      ctx.close();
-                    }
-                  }
-                }
-              });
-              pipeline.addLast(commandHandler);
-            }
-          });
+  public Ingester(ChannelInitializer initializer, int port) {
+    this.listeningPort = port;
+    this.initializer = initializer;
+  }
 
-      // Start the server.
-      ChannelFuture f = b.bind().sync();
+  /**
+   * Creates the ChannelInitializer for this ingester
+   */
+  private void createInitializer(@Nullable final List<Function<Channel, ChannelHandler>> decoders, final ChannelHandler commandHandler) {
+    this.initializer = new ChannelInitializer<SocketChannel>() {
+      @Override
+      public void initChannel(SocketChannel ch) throws Exception {
+        ChannelPipeline pipeline = ch.pipeline();
+        addDecoders(ch, decoders);
+        addIdleTimeoutHandler(pipeline);
+        pipeline.addLast(commandHandler);
+      }
+      };
+  }
 
-      // Wait until the server socket is closed.
-      f.channel().closeFuture().sync();
-    } catch (InterruptedException e) {
-      // Server was interrupted
-      e.printStackTrace();
+  /**
+   * Adds an idle timeout handler to the given pipeline
+   * @param pipeline the pipeline to add the idle timeout handler
+   */
+  protected void addIdleTimeoutHandler(final ChannelPipeline pipeline) {
+    // Shared across all reports for proper batching
+    pipeline.addLast("idleStateHandler",
+                     new IdleStateHandler(CHANNEL_IDLE_TIMEOUT_IN_SECS_DEFAULT,
```

Comment:
method invocations do not align like this, check your IDE settings (they should just be 4-character indented)